boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs

```diff
@@ -0,0 +1,78 @@
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    syntax::{
+        ast::{node::AsyncFunctionExpr, Keyword, Punctuator},
+        lexer::TokenKind,
+        parser::{
+            function::{FormalParameters, FunctionBody},
+            statement::BindingIdentifier,
+            AllowYield, Cursor, ParseError, TokenParser,
+        },
+    },
+    BoaProfiler,
+};
+
+use std::io::Read;
+
+/// Async Function expression parsing.
+///
+/// More information:
+///  - [MDN documentation][mdn]
+///  - [ECMAScript specification][spec]
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/async_function
+/// [spec]: https://www.ecma-international.org/ecma-262/11.0/index.html#prod-AsyncFunctionExpression
+#[derive(Debug, Clone, Copy)]
+pub(super) struct AsyncFunctionExpression {
+    allow_yield: AllowYield,
+}
+
+impl AsyncFunctionExpression {
+    /// Creates a new `AsyncFunctionExpression` parser.
+    pub(super) fn new<Y>(allow_yield: Y) -> Self
+    where
+        Y: Into<AllowYield>,
+    {
+        Self {
+            allow_yield: allow_yield.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for AsyncFunctionExpression
+where
+    R: Read,
+{
+    type Output = AsyncFunctionExpr;
+
+    fn parse(self, cursor: &mut Cursor<R>) -> Result<Self::Output, ParseError> {
+        let _timer = BoaProfiler::global().start_event("AsyncFunctionExpression", "Parsing");
+        cursor.expect_no_skip_lineterminator(Keyword::Function, "async function expression")?;
+
+        let tok = cursor.peek(0)?;
+
+        let name = if let Some(token) = tok {
+            match token.kind() {
+                TokenKind::Punctuator(Punctuator::OpenParen) => None,
+                _ => Some(BindingIdentifier::new(self.allow_yield, true).parse(cursor)?),
+            }
+        } else {
+            return Err(ParseError::AbruptEnd);
+        };
+
+        cursor.expect(Punctuator::OpenParen, "function expression")?;
+
+        let params = FormalParameters::new(!self.allow_yield.0, true).parse(cursor)?;
```

Comment:
Doesn't ~yield mean NOT yield rather than false?

Reply:
As far as I understand, there are 3 options, either, we pass the same value with `?`, we pass `false` with `~` or we pass `true` with `+`. Maybe @jasonwilliams or @HalidOdat can confirm this.

Reply:
To be honest even after reading the spec (https://tc39.es/ecma262/#sec-grammar-notation) I'm still unsure - I took their usage of 'not' to mean logical inverse but perhaps they meant logical false

Reply:
Looking at this section of the spec, if it's `+`, then, we only take into account when that parameter is set (`true` in Boa), when it's `?`, then, it will depend on that parameter being `true` or `false` in the current production, so we just pass it, and if it's `~`, then, we only take into account when that parameter is not set, so `false` in Boa:
![Screenshot from 2020-10-12 13-26-21](https://user-images.githubusercontent.com/597469/95741449-933bef00-0c8e-11eb-96e7-df9588f48aa3.png)

Reply:
Hmm ok